packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -3253,6 +3270,68 @@ export class ContainerRuntime
 		return result;
 	}
 
+	// eslint-disable-next-line import/no-deprecated
+	private stageControls: StageControlsExperimental | undefined;
+
+	/**
+	 * If true, the ContainerRuntime is not submitting any new ops to the ordering service.
+	 * Ops submitted to the ContainerRuntime while in Staging Mode will be queued in the PendingStateManager,
+	 * either to be discarded or committed later (via the Stage Controls returned from enterStagingMode).
+	 */
+	public get inStagingMode(): boolean {
+		return this.stageControls !== undefined;
+	}
+
+	/**
+	 * Enter Staging Mode, such that ops submitted to the ContainerRuntime will not be sent to the ordering service.
+	 * To exit Staging Mode, call either discardChanges or commitChanges on the Stage Controls returned from this method.
+	 *
+	 * @returns StageControlsExperimental - Controls for exiting Staging Mode.
+	 */
+	// eslint-disable-next-line import/no-deprecated
+	public enterStagingMode = (): StageControlsExperimental => {
+		if (this.stageControls !== undefined) {
+			throw new Error("already in staging mode");
+		}
+
+		// Make sure all BatchManagers are empty before entering staging mode,
+		// since we mark whole batches as "staged" or not to indicate whether to submit them.
+		this.outbox.flush();
+		const exitStagingMode = (discardOrCommit: () => void) => (): void => {
+			// Final flush of any last staged changes
+			//* NIT: Don't need to pass true.  And then these two params can be joined into an atomic object?
+			this.outbox.flush(undefined, true /* staged */);
+
+			this.stageControls = undefined;
```

Comment:
(I know I'm commenting on the draft/WIP PR, so feel free to resolve this wherever makes sense) nit: seems better to clear this after invoking `discardOrCommit()`, maybe even in a try-finally.

Reply:
(This is actually very close to final)

Have to clear it first so the ops submitted during resubmit during commit don't get marked as staged

Reply:
we can continue to iterate here once the core is checked in. i've been refraining from doing to much work in this branch to avoid any thrashing while we move to main.

Reply:
ah yeah of course